### ISSUE TITLE
Definitely bugs somewhere

### ISSUE THREAD

**Issue by chungyang** [jadore801120/attention-is-all-you-need-pytorch#101]
I trained on only 10 sentences to see if the model can overfit it and I also used a validation set of 7883 sentences. The data is IWST2016 de-en by the way. The result I got during training indicates bugs in the implementation. 

[ Epoch 0 ]
  - (Training)   ppl:  173.38731, accuracy: 0.939 %, elapse: 0.004 min
  - (Validation) ppl:  177.59718, accuracy: 1.775 %, elapse: 0.197 min
    - [Info] The checkpoint file has been updated.
[ Epoch 1 ]
  - (Training)   ppl:  181.91788, accuracy: 1.878 %, elapse: 0.004 min
  - (Validation) ppl:  176.28372, accuracy: 1.917 %, elapse: 0.196 min
    - [Info] The checkpoint file has been updated.
[ Epoch 2 ]
  - (Training)   ppl:  163.84926, accuracy: 1.878 %, elapse: 0.003 min
  - (Validation) ppl:  174.44017, accuracy: 2.167 %, elapse: 0.197 min
    - [Info] The checkpoint file has been updated.
[ Epoch 3 ]
  - (Training)   ppl:  170.36798, accuracy: 0.939 %, elapse: 0.003 min
  - (Validation) ppl:  172.20245, accuracy: 2.431 %, elapse: 0.197 min
    - [Info] The checkpoint file has been updated.
[ Epoch 4 ]
  - (Training)   ppl:  174.74831, accuracy: 1.408 %, elapse: 0.003 min
  - (Validation) ppl:  169.75132, accuracy: 2.709 %, elapse: 0.197 min
    - [Info] The checkpoint file has been updated.
[ Epoch 5 ]
  - (Training)   ppl:  162.07248, accuracy: 2.347 %, elapse: 0.003 min
  - (Validation) ppl:  167.20357, accuracy: 3.028 %, elapse: 0.197 min
    - [Info] The checkpoint file has been updated.
[ Epoch 6 ]
  - (Training)   ppl:  135.76818, accuracy: 1.408 %, elapse: 0.003 min
  - (Validation) ppl:  164.84782, accuracy: 3.287 %, elapse: 0.197 min
    - [Info] The checkpoint file has been updated.
[ Epoch 7 ]
  - (Training)   ppl:  153.03017, accuracy: 2.347 %, elapse: 0.003 min
  - (Validation) ppl:  162.73445, accuracy: 3.484 %, elapse: 0.197 min
    - [Info] The checkpoint file has been updated.
[ Epoch 8 ]
  - (Training)   ppl:  144.36503, accuracy: 2.817 %, elapse: 0.003 min
  - (Validation) ppl:  160.63012, accuracy: 3.599 %, elapse: 0.196 min
    - [Info] The checkpoint file has been updated.
[ Epoch 9 ]
  - (Training)   ppl:  132.34567, accuracy: 6.573 %, elapse: 0.003 min
  - (Validation) ppl:  158.51897, accuracy: 3.674 %, elapse: 0.196 min
    - [Info] The checkpoint file has been updated.
[ Epoch 10 ]
  - (Training)   ppl:  130.91847, accuracy: 6.103 %, elapse: 0.003 min
  - (Validation) ppl:  156.63602, accuracy: 3.720 %, elapse: 0.196 min
    - [Info] The checkpoint file has been updated.
[ Epoch 11 ]
  - (Training)   ppl:  120.60263, accuracy: 6.103 %, elapse: 0.003 min
  - (Validation) ppl:  154.87947, accuracy: 3.743 %, elapse: 0.196 min
    - [Info] The checkpoint file has been updated.
[ Epoch 12 ]
  - (Training)   ppl:  118.87778, accuracy: 6.103 %, elapse: 0.003 min
  - (Validation) ppl:  153.55154, accuracy: 3.751 %, elapse: 0.196 min
    - [Info] The checkpoint file has been updated.
[ Epoch 13 ]
  - (Training)   ppl:  116.07172, accuracy: 5.634 %, elapse: 0.003 min
  - (Validation) ppl:  152.69443, accuracy: 3.764 %, elapse: 0.196 min
    - [Info] The checkpoint file has been updated.
[ Epoch 14 ]
  - (Training)   ppl:  110.05531, accuracy: 7.512 %, elapse: 0.003 min
  - (Validation) ppl:  151.97540, accuracy: 3.760 %, elapse: 0.196 min
[ Epoch 15 ]
  - (Training)   ppl:  96.67249, accuracy: 7.981 %, elapse: 0.003 min
  - (Validation) ppl:  151.73769, accuracy: 3.758 %, elapse: 0.197 min
[ Epoch 16 ]
  - (Training)   ppl:  103.35449, accuracy: 7.512 %, elapse: 0.003 min
  - (Validation) ppl:  152.11966, accuracy: 3.747 %, elapse: 0.196 min
[ Epoch 17 ]
  - (Training)   ppl:  99.91647, accuracy: 7.512 %, elapse: 0.003 min
  - (Validation) ppl:  152.85609, accuracy: 3.738 %, elapse: 0.197 min
[ Epoch 18 ]
  - (Training)   ppl:  95.96763, accuracy: 7.512 %, elapse: 0.003 min
  - (Validation) ppl:  154.42918, accuracy: 3.734 %, elapse: 0.196 min
[ Epoch 19 ]
  - (Training)   ppl:  87.69083, accuracy: 6.573 %, elapse: 0.003 min
  - (Validation) ppl:  156.56630, accuracy: 3.728 %, elapse: 0.197 min

The model clearly has problem overfit the training set with just 10 sentences and that clearly shows the implementation has bugs somewhere.

**Comment by sanyam5** [comment]
What hyperparameters did you use? Optimizer, learning rate, layer sizes, etc. Did you also try different hyperparameters? Can you post your training script?

**Comment by chungyang** [comment]
These are the arguments I used. Beside epoch number, pretty much everything is the same as when you clone the project. The training script is exactly the same. I have not modified it. The only thing that's different is that I used tokenized IWST2016 de-en dataset. 

```
    parser.add_argument('-epoch', type=int, default=20)
    parser.add_argument('-batch_size', type=int, default=64)

    #parser.add_argument('-d_word_vec', type=int, default=512)
    parser.add_argument('-d_model', type=int, default=512)
    parser.add_argument('-d_inner_hid', type=int, default=2048)
    parser.add_argument('-d_k', type=int, default=64)
    parser.add_argument('-d_v', type=int, default=64)

    parser.add_argument('-n_head', type=int, default=8)
    parser.add_argument('-n_layers', type=int, default=6)
    parser.add_argument('-n_warmup_steps', type=int, default=4000)

    parser.add_argument('-dropout', type=float, default=0.1)
```

I could certainly try more iterations, but I'm not sure if this is simply a convergence issue as the training set only has 10 sentences. Even with a single layer LSTM encoder decoder architecture is able to overfit the data pretty fast.

**Comment by sanyam5** [comment]
According to the optimization script https://github.com/jadore801120/attention-is-all-you-need-pytorch/blob/master/transformer/Optim.py it seems like the initial learning rate is very high. I think your issue is one of convergence. I'd suggest letting the model train for longer.
